spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java

```diff
@@ -285,4 +300,87 @@ protected void set(InternalRow struct, int pos, Object value) {
       }
     }
   }
+
+  private static class ComplexUnionReader implements ValueReader<InternalRow> {
+    private static final String UNION_TAG_FIELD_NAME = "tag";
+    private final Schema schema;
+    private final List<Schema> branches;
+    private final ValueReader[] readers;
+    private int nullIndex;
+    private final int[] projectedFieldIdsToIdxInReturnedRow;
+    private boolean isTagFieldProjected;
+    private int numOfFieldsInReturnedRow;
+    private int nullTypeIndex;
+
+    private ComplexUnionReader(Schema schema, List<ValueReader<?>> readers, Type expected) {
+      this.schema = schema;
+      this.branches = schema.getTypes();
+      this.readers = new ValueReader[readers.size()];
+      for (int i = 0; i < this.readers.length; i += 1) {
+        this.readers[i] = readers.get(i);
+      }
+
+      // checking if NULL type exists in Avro union schema
+      this.nullTypeIndex = -1;
+      for (int i = 0; i < this.schema.getTypes().size(); i++) {
+        Schema alt = this.schema.getTypes().get(i);
+        if (Objects.equals(alt.getType(), Schema.Type.NULL)) {
+          this.nullTypeIndex = i;
+          break;
+        }
+      }
+
+      // Creating an integer array to track the mapping between the index of fields to be projected
+      // and the index of the value for the field stored in the returned row,
+      // if the value for a field equals to -1, it means the value of this field should not be
+      // stored
+      // in the returned row
+      int numberOfTypes =
+          this.nullTypeIndex == -1
+              ? this.schema.getTypes().size()
+              : this.schema.getTypes().size() - 1;
+      this.projectedFieldIdsToIdxInReturnedRow = new int[numberOfTypes];
+      Arrays.fill(this.projectedFieldIdsToIdxInReturnedRow, -1);
+      this.numOfFieldsInReturnedRow = 0;
+      this.isTagFieldProjected = false;
+      for (Types.NestedField expectedStructField : expected.asStructType().fields()) {
+        String fieldName = expectedStructField.name();
+        if (fieldName.equals(UNION_TAG_FIELD_NAME)) {
+          this.isTagFieldProjected = true;
+          this.numOfFieldsInReturnedRow++;
+          continue;
+        }
+        int projectedFieldIndex = Integer.valueOf(fieldName.substring(5));
+        this.projectedFieldIdsToIdxInReturnedRow[projectedFieldIndex] =
+            this.numOfFieldsInReturnedRow++;
+      }
+    }
+
+    @Override
+    public InternalRow read(Decoder decoder, Object reuse) throws IOException {
```

Comment:
Should not the logic here be:
* Iterate on the Avro schema. For each branch, get the field ID from the Avro schema annotation. 
* The assumption is Avro schema union preserves all the union branches even if some are not projected. So we still need to figure out if a field is projected or not. This can happen by checking the field ID from the step above in the expected Iceberg schema. If the field ID is projected, populate the InternalRow index using the next suitable reader (hopefully reader order is preset properly in `AvroSchemaWithTypeVisitor` to match the expected projection).
* If the field ID is not projected, skip.
The above logic can be split/refactored between the constructor and the `read` method for efficiency.
@rdblue Let me know if this matches your understanding.

Reply:
I agree with @wmoustafa, although I think that this is correct to make the mapping array. That way the implementation is straightforward:

```java
  InternalRow row = reuseOrCreate(reuse); // this is where setNullAt happens
  int index = decoder.readIndex();
  int destIndex = projectionIndexes[index];
  if (destIndex >= 0) {
    Object value = readers[index].read(decoder, get(reuse, destIndex));
    row.update(destIndex, value);
  } else {
    readers[index].read(decoder, null);
  }

  return row;
```

Reply:
Per Ryan's suggestion, Avro Schema is not passed into ComplexUnionReader.  The approach of mapping array is used to track the relationship between a branch type and the position of its value in the returned row.